src/Microsoft.Identity.Web.DownstreamApi/DownstreamApi.cs

```diff
@@ -302,43 +303,65 @@ internal partial class DownstreamApi : IDownstreamApi
+            using HttpRequestMessage httpRequestMessage = new HttpRequestMessage(
+                new HttpMethod(effectiveOptions.HttpMethod),
                 apiUrl);
-            if (content != null)
-            {
-                httpRequestMessage.Content = content;
-            }
 
-            // Obtention of the authorization header (except when calling an anonymous endpoint
-            // which is done by not specifying any scopes
-            if (effectiveOptions.Scopes != null && effectiveOptions.Scopes.Any())
-            {
-                string authorizationHeader = appToken ?
-                    await _authorizationHeaderProvider.CreateAuthorizationHeaderForAppAsync(
-                                            effectiveOptions.Scopes.FirstOrDefault()!,
-                                            effectiveOptions,
-                                            cancellationToken).ConfigureAwait(false) :
-                    await _authorizationHeaderProvider.CreateAuthorizationHeaderForUserAsync(
-                                            effectiveOptions.Scopes,
-                                            effectiveOptions,
-                                            user,
-                                            cancellationToken).ConfigureAwait(false);
-                httpRequestMessage.Headers.Add(Authorization, authorizationHeader);
-            }
-            else
-            {
-                Logger.UnauthenticatedApiCall(_logger, null);
-            }
-            // Opportunity to change the request message
-            effectiveOptions.CustomizeHttpRequestMessage?.Invoke(httpRequestMessage);
-
-            // Send the HTTP message
-            using HttpClient client = string.IsNullOrEmpty(serviceName) ? _httpClientFactory.CreateClient() : _httpClientFactory.CreateClient(serviceName);
-            return await client.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+            await UpdateRequestAsync(httpRequestMessage);
+
+            // Send the HTTP message           
+            var downstreamApiResult = await client.SendAsync(httpRequestMessage, cancellationToken).ConfigureAwait(false);
+
+            if (downstreamApiResult.StatusCode != System.Net.HttpStatusCode.Unauthorized)
+            {
+                return downstreamApiResult;
+            }
+
+            // Retry if the resource send 401 Unauthorized with WWW-Authenticate header and claims
+            effectiveOptions.AcquireTokenOptions.Claims = WwwAuthenticateParameters.GetClaimChallengeFromResponseHeaders(downstreamApiResult.Headers);
+
+            using HttpRequestMessage retryHttpRequestMessage = new(
+                new HttpMethod(effectiveOptions.HttpMethod),
+                apiUrl);
+
+            await UpdateRequestAsync(retryHttpRequestMessage);
+
+            return await client.SendAsync(retryHttpRequestMessage, cancellationToken).ConfigureAwait(false);
         }
```

Comment:
I think we need the e2e test before merging.

Reply:
Tested E2E manually. For integration tests, we'll need to wait for the lab to have a CAE test app set up then..

Reply:
Where can we get details on the E2E test that was done?

Reply:
In the PR description. Which details are you looking for?

Reply:
I tried testing this several several months ago and couldn't get it to work. I guess I will need more steps on what exactly to do.

Reply:
Added more testing steps in the PR description.

Reply:
Is this the only blocking comment remaining? Integration testing?

Reply:
I've reduced the scope of this PR. It doesn't set CAE flag by default; it only retries downstream API requests if the first response was 401 and contains claims. Added more unit tests.